dbms/src/Common/SharedMutexProtected.h

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/nocopyable.h>
+#include <common/defines.h>
+#include <common/types.h>
+
+#include <shared_mutex>
+
+namespace DB
+{
+
+template <typename T>
+class SharedMutexProtected
+{
+    DISALLOW_COPY_AND_MOVE(SharedMutexProtected);
+
+private:
+    template <typename U, typename Lock>
+    class Locked
+    {
+        DISALLOW_COPY_AND_MOVE(Locked);
```

Comment:
Why can't it move?

Reply:
`U` maybe non-moveable

Reply:
U is not movable, a lock to U could be movable though IMO?

Reply:
But it is not only a lock, It all contains the reference.

Reply:
What's the difference then? References and pointers can be safely elsewhere.

Reply:
```cpp
{
    SharedMutexProtected<> mp;
    auto lock = mp.lockShared();
    do_something;
    auto task = [std::move(lock)] () { // seems reasonable, because the lock can not be null in task.
       lock->xxx
   }
   add_task_to_thread_pool(task);
}

// thread_pool execute task here.
```

Reply:
Yes, this is one case. However, IMO, preventing moving doesn't ensure the lock always refer to a valid pointer

Reply:
The `value` can not be accessed. We only need to make sure return `Lock` is valid.

Reply:
> The value can not be accessed
I don't get that. I find `operator->` can access `value` in `Locked`. So I think the Locked should not outlive `value` which it is protected.